### ISSUE TITLE
ninja build fails for release due to Werror flag (debug works well)

### ISSUE THREAD

**Issue by guy4261** [apple/turicreate#1667]
I am building `turicreate`:
* git commit sha1 is `6942101316b67b394188732bf5ee9558e70d6db9`
* on a Docker container - `centos:6.9`
* using ninja

After a successful `./configure`, `debug` finished building well, while `release` dies. Seems like it's due to the `Werror` flag:

```bash
FAILED: test/fileio/CMakeFiles/fixed_size_cache_manager_test.cxxtest.dir/fixed_size_cache_manager_test.cxx.o
/turicreate/deps/local/bin/cxx  -DBOOST_FUSION_INVOKE_FUNCTION_OBJECT_MAX_ARITY=12 -DBOOST_FUSION_INVOKE_MAX_ARITY=12 -DBOOST_FUSION_INVOKE_PROCEDURE_MAX_ARITY=12 -DBOOST_RESULT_OF_USE_TR1_WITH_DECLTYPE_FALLBACK -DBOOST_SPIRIT_THREAD_SAFE -DCOMPILER_HAS_NOEXCEPT_WHAT_ON_EXCEPTIONS -DCURL_STATICLIB -DEIGEN_MPL2_ONLY -DFUSION_MAX_VECTOR_SIZE=20 -DHAS_BOOST -DHAS_CURL -DHAS_LIBJPEG -DHAS_LIBPNG -DHAS_LIBZ -DHAS_NANOMSG -DHAS_OPENSSL -DIN_TURI_SOURCE_TREE -DNN_STATIC_LIB -DTC_BUILD_PATH_BASE=\"/turicreate\" -DTC_BUILD_VISUALIZATION_CLIENT -DTC_ENABLE_REMOTEFS -DTC_EXTRA_OPTIMIZATION_FLAGS -DTC_HAS_PYTHON -DTC_PLATFORM_LINUX -D__NO_OPENMP__ -I../src/external/coremltools_wrap/coremltools -I../deps/local/include/libxml2 -I../src -I../src/platform -isystem ../deps/local/include -isystem ../src/external -isystem ../src/external/aws-sdk-cpp/aws-cpp-sdk-core/include -isystem ../src/external/aws-sdk-cpp/aws-cpp-sdk-s3/include -isystem ../src/external/google -isystem src -O3 -DNDEBUG  -fpeel-loops -funswitch-loops -ftracer  -Wno-unused-local-typedefs  -Wno-attributes -march=nehalem -mtune=generic -Wall -g -Wno-enum-compare -Wno-conversion-null -ftemplate-depth=900 -Wno-unknown-pragmas  -pthread -Werror -DHAVE__int128_t -DHASH_FOR_INT128_DEFINED -DHAVE__uint128_t -DHASH_FOR_UINT128_DEFINED  -DNDEBUG -DHAVE_PTHREAD=1 -fPIE   -Dgoogle=__tc_google -DCoreML=__tc_CoreML -DHAVE_PTHREAD=1 -std=gnu++11 -MD -MT test/fileio/CMakeFiles/fixed_size_cache_manager_test.cxxtest.dir/fixed_size_cache_manager_test.cxx.o -MF test/fileio/CMakeFiles/fixed_size_cache_manager_test.cxxtest.dir/fixed_size_cache_manager_test.cxx.o.d -o test/fileio/CMakeFiles/fixed_size_cache_manager_test.cxxtest.dir/fixed_size_cache_manager_test.cxx.o -c ../test/fileio/fixed_size_cache_manager_test.cxx
In file included from ../test/fileio/fixed_size_cache_manager_test.cxx:5:0:
../src/fileio/fixed_size_cache_manager.hpp: In member function ‘void fixed_size_cache_manager_test::test_new_cache()’:
../src/fileio/fixed_size_cache_manager.hpp:110:13: error: ‘dummy’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
       memcpy(data + size, c, bufsize);
       ~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~
cc1plus: all warnings being treated as errors
```

**Comment by guy4261** [comment]
Replacing `-Werror` with `-Wno-error`, the release ninja build finishes well.

```patch
[root@75564d656c6d turicreate]# git diff CMakeLists.txt
diff --git a/CMakeLists.txt b/CMakeLists.txt
index cd624a0..c6b5f38 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -344,11 +344,11 @@ if (CLANG)

   set(CPP11_FLAGS "-stdlib=libc++ -Wno-deprecated-register -Wno-enum-compare -Wno-conversion-null -Wno-constant-logical-operand -

-  set(WERROR_FLAGS "-Werror -Wall -Wsign-compare")
+  set(WERROR_FLAGS "-Wno-error -Wall -Wsign-compare")
 else()
   # non-clang
   set(CPP11_FLAGS "-Wno-enum-compare -Wno-conversion-null -ftemplate-depth=900 -Wno-unknown-pragmas")
-  set(WERROR_FLAGS "-Werror")
+  set(WERROR_FLAGS "-Wno-error")
 endif()
```

**Comment by hoytak** [comment]
This is a legit compiler warning -- catching these is exactly why we have -Werror turned on in the first place :-).

**Comment by guy4261** [comment]
Might be a good idea to fix the code then, isn't it? This is the *only* error this `Werror` raises   https://github.com/apple/turicreate/issues/1667#issue-425339365

```make
FAILED: test/fileio/CMakeFiles/fixed_size_cache_manager_test.cxxtest.dir/fixed_size_cache_manager_test.cxx.o
/turicreate/deps/local/bin/cxx  -DBOOST_FUSION_INVOKE_FUNCTION_OBJECT_MAX_ARITY=12 -DBOOST_FUSION_INVOKE_MAX_ARITY=12 -DBOOST_FUSION_INVOKE_PROCEDURE_MAX_ARITY=12 -DBOOST_RESULT_OF_USE_TR1_WITH_DECLTYPE_FALLBACK -DBOOST_SPIRIT_THREAD_SAFE -DCOMPILER_HAS_NOEXCEPT_WHAT_ON_EXCEPTIONS -DCURL_STATICLIB -DEIGEN_MPL2_ONLY -DFUSION_MAX_VECTOR_SIZE=20 -DHAS_BOOST -DHAS_CURL -DHAS_LIBJPEG -DHAS_LIBPNG -DHAS_LIBZ -DHAS_NANOMSG -DHAS_OPENSSL -DIN_TURI_SOURCE_TREE -DNN_STATIC_LIB -DTC_BUILD_PATH_BASE=\"/turicreate\" -DTC_BUILD_VISUALIZATION_CLIENT -DTC_ENABLE_REMOTEFS -DTC_EXTRA_OPTIMIZATION_FLAGS -DTC_HAS_PYTHON -DTC_PLATFORM_LINUX -D__NO_OPENMP__ -I../src/external/coremltools_wrap/coremltools -I../deps/local/include/libxml2 -I../src -I../src/platform -isystem ../deps/local/include -isystem ../src/external -isystem ../src/external/aws-sdk-cpp/aws-cpp-sdk-core/include -isystem ../src/external/aws-sdk-cpp/aws-cpp-sdk-s3/include -isystem ../src/external/google -isystem src -O3 -DNDEBUG  -fpeel-loops -funswitch-loops -ftracer  -Wno-unused-local-typedefs  -Wno-attributes -march=nehalem -mtune=generic -Wall -g -Wno-enum-compare -Wno-conversion-null -ftemplate-depth=900 -Wno-unknown-pragmas  -pthread -Werror -DHAVE__int128_t -DHASH_FOR_INT128_DEFINED -DHAVE__uint128_t -DHASH_FOR_UINT128_DEFINED  -DNDEBUG -DHAVE_PTHREAD=1 -fPIE   -Dgoogle=__tc_google -DCoreML=__tc_CoreML -DHAVE_PTHREAD=1 -std=gnu++11 -MD -MT test/fileio/CMakeFiles/fixed_size_cache_manager_test.cxxtest.dir/fixed_size_cache_manager_test.cxx.o -MF test/fileio/CMakeFiles/fixed_size_cache_manager_test.cxxtest.dir/fixed_size_cache_manager_test.cxx.o.d -o test/fileio/CMakeFiles/fixed_size_cache_manager_test.cxxtest.dir/fixed_size_cache_manager_test.cxx.o -c ../test/fileio/fixed_size_cache_manager_test.cxx
In file included from ../test/fileio/fixed_size_cache_manager_test.cxx:5:0:
../src/fileio/fixed_size_cache_manager.hpp: In member function ‘void fixed_size_cache_manager_test::test_new_cache()’:
../src/fileio/fixed_size_cache_manager.hpp:110:13: error: ‘dummy’ may be used uninitialized in this function [-Werror=maybe-uninitialized]
       memcpy(data + size, c, bufsize);
```

**Comment by nickjong** [comment]
I'm assuming Hoyt just closed this prematurely, pending #1690
